zkevm-circuits/src/evm_circuit/execution/sha3.rs

```diff
@@ -0,0 +1,201 @@
+use bus_mapping::{circuit_input_builder::CopyDataType, evm::OpcodeId};
+use eth_types::{evm_types::GasCost, Field, ToLittleEndian};
+use gadgets::util::Expr;
+use halo2_proofs::plonk::Error;
+
+use crate::evm_circuit::{
+    param::N_BYTES_MEMORY_WORD_SIZE,
+    step::ExecutionState,
+    util::{
+        common_gadget::SameContextGadget,
+        constraint_builder::{ConstraintBuilder, StepStateTransition, Transition},
+        memory_gadget::{MemoryAddressGadget, MemoryCopierGasGadget, MemoryExpansionGadget},
+        rlc, CachedRegion, Cell, Word,
+    },
+    witness::{Block, Call, ExecStep, Transaction},
+};
+
+use super::ExecutionGadget;
+
+#[derive(Clone, Debug)]
+pub(crate) struct Sha3Gadget<F> {
+    same_context: SameContextGadget<F>,
+    memory_address: MemoryAddressGadget<F>,
+    sha3_rlc: Word<F>,
+    copy_rwc_inc: Cell<F>,
+    rlc_acc: Cell<F>,
+    memory_expansion: MemoryExpansionGadget<F, 1, N_BYTES_MEMORY_WORD_SIZE>,
+    memory_copier_gas: MemoryCopierGasGadget<F, { GasCost::COPY_SHA3 }>,
+}
+
+impl<F: Field> ExecutionGadget<F> for Sha3Gadget<F> {
+    const EXECUTION_STATE: ExecutionState = ExecutionState::SHA3;
+
+    const NAME: &'static str = "SHA3";
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
+
+        let offset = cb.query_cell();
+        let size = cb.query_rlc();
+        let sha3_rlc = cb.query_word();
```

Comment:
just curious: query_word is same as query_rlc. So line 40 and line 41 is there any indication by using `query_rlc` and `query_word`?

Reply:
There is no such difference. I used `query_word` for SHA3's RLC value because it is a 32-bytes value pushed to the stack, and I used `query_rlc` for the size because even though its coming from the stack, the `MemoryAddressGadget` later constraints it to be within 5-bytes. So the differentiation is simply for me to note that one is 32-bytes while the other is not. I could use `query_rlc` for `sha3_rlc` and it shouldn't be a problem :)